internal/peertimeoutmanager/peertimeoutmanager.go

```diff
@@ -0,0 +1,201 @@
+package peertimeoutmanager
+
+import (
+	"context"
+	"time"
+
+	peer "github.com/libp2p/go-libp2p-core/peer"
+
+	bstimer "github.com/ipfs/go-bitswap/internal/timer"
+)
+
+const (
+	// The time to allow a peer to respond before assuming it is unresponsive.
+	peerResponseTimeout = 10 * time.Second
+
+	// We keep track of requests in a map, and the order of requests in an
+	// array. As responses come in items are deleted from the map and set to
+	// inactive in the array. If the array and map get too far out of sync
+	// we pause to filter the array. requestGCCount is the maximum allowed
+	// difference between map and array length before forcing a filtering.
+	requestGCCount = 256
+)
+
+type request struct {
+	// The peer the request was sent to
+	p peer.ID
+	// The time at which the request was sent
+	sent time.Time
+	// active is true until a response is received
+	active bool
+}
+
+// FnPeerResponseTimeout is called when one or more peers time out
+type FnPeerResponseTimeout func([]peer.ID)
+
+// PeerTimeoutManager keeps track of requests and responses to peers.
+// It reports a timeout if there is no response from a peer to any request
+// for some period of time.
+type PeerTimeoutManager struct {
+	ctx                   context.Context
+	peerResponseTimeout   time.Duration
+	onPeerResponseTimeout FnPeerResponseTimeout
+	// queue of requests that are sent
+	requestSent chan request
+	// queue of responding peers
+	responseReceived chan peer.ID
+}
+
+// New creates a new PeerTimeoutManager.
+func New(ctx context.Context, onTimeout FnPeerResponseTimeout) *PeerTimeoutManager {
+	return newPeerTimeoutManager(ctx, onTimeout, peerResponseTimeout)
+}
+
+// Used by the tests
+func newPeerTimeoutManager(ctx context.Context, onTimeout FnPeerResponseTimeout, timeout time.Duration) *PeerTimeoutManager {
+	ptm := &PeerTimeoutManager{
+		ctx:                   ctx,
+		peerResponseTimeout:   timeout,
+		onPeerResponseTimeout: onTimeout,
+		requestSent:           make(chan request, 128),
+		responseReceived:      make(chan peer.ID, 128),
+	}
+
+	go ptm.run()
+
+	return ptm
+}
+
+// Called when a request is sent to a peer
+func (ptm *PeerTimeoutManager) RequestSent(p peer.ID) {
+	select {
+	case ptm.requestSent <- request{p: p, sent: time.Now(), active: true}:
+	case <-ptm.ctx.Done():
+	}
+}
+
+// Called when a response is received from a peer
+func (ptm *PeerTimeoutManager) ResponseReceived(p peer.ID) {
+	select {
+	case ptm.responseReceived <- p:
+	case <-ptm.ctx.Done():
+	}
+}
+
+// The main run loop for the class
+func (ptm *PeerTimeoutManager) run() {
+	// Create a timer to detect when a peer doesn't respond for too long
+	responseTimer := time.NewTimer(0)
+	defer responseTimer.Stop()
+	bstimer.StopTimer(responseTimer)
+
+	// Keep track of requests sent to a peer, and the order of requests
+	requestsSent := make(map[peer.ID]*request)
+	var inOrder []*request
+	for {
```

Comment:
Raised OOB in a call: We need to handle cancels. To do that, we're probably going to have to track every request.